src/filters/filters.ts

```diff
@@ -0,0 +1,233 @@
+import { linearRegressionLine, linearRegression } from "simple-statistics";
+import { timeUnits } from "../duration/duration";
+import { StatisticValue } from "../recorder-types";
+import { HassEntity, YValue } from "../types";
+
+const castFloat = (y: any) => parseFloat(y);
+const myEval = typeof window != "undefined" ? window.eval : global.eval;
+
+type FilterData = {
+  xs: Date[];
+  ys: YValue[];
+  states: HassEntity[];
+  statistics: StatisticValue[];
+  meta: HassEntity["attributes"];
+  vars: Record<any, any>;
+};
+export type FilterFn = (p: FilterData) => Partial<FilterData>;
+
+const mapNumbers = (ys: YValue[], fn: (y: number, i: number) => number) =>
+  ys.map((y, i) => {
+    const n = castFloat(y);
+    if (Number.isNaN(n)) return y;
+    return fn(n, i);
+  });
+
+/**
+ * Removes from all params the indexes for which ys is not numeric, and parses ys to numbers.
+ * WARNING: when used inside a filter, it is important to return all arrays. Otherwise the lengths
+ * between say ys and states won't be consistent
+ */
+const force_numeric: (p: FilterData) => { ys: number[] } & FilterData = ({
+  xs,
+  ys: ys2,
+  states,
+  statistics,
+  ...rest
+}) => {
+  const ys = ys2.map((y) => castFloat(y));
+  const mask = ys.map((y) => !isNaN(y));
+  return {
+    ys: ys.filter((_, i) => mask[i]),
+    xs: xs.filter((_, i) => mask[i]),
+    states: states.filter((_, i) => mask[i]),
+    statistics: statistics.filter((_, i) => mask[i]),
+    ...rest,
+  };
+};
+
+const filters = {
+  force_numeric: () => force_numeric,
+  add:
+    (val: number) =>
+    ({ ys }) => ({
+      ys: mapNumbers(ys, (y) => y + val),
+    }),
+  multiply:
+    (val: number) =>
+    ({ ys }) => ({
+      ys: mapNumbers(ys, (y) => y * val),
+    }),
+  calibrate_linear:
+    (mappingStr: `${number} -> ${number}`[]) =>
+    ({ ys, meta }) => {
+      const mapping = mappingStr.map((str) => str.split("->").map(parseFloat));
+      const regression = linearRegression(mapping);
+      const mapper = linearRegressionLine(regression);
+      return {
+        ys: mapNumbers(ys, mapper),
+        meta: { ...meta, regression },
+      };
+    },
+  derivate:
+    (unit: keyof typeof timeUnits = "h") =>
+    ({ xs, ys, meta }) => {
+      const last = {
+        x: +xs[0],
+        y: NaN,
+      };
+      return {
+        meta: {
+          unit_of_measurement: `${meta.unit_of_measurement}/${unit}`,
+        },
+        xs,
+        ys: mapNumbers(ys, (y, i) => {
+          const x = +xs[i];
+          const dateDelta = (x - last.x) / timeUnits[unit];
+          const yDelta = (y - last.y) / dateDelta;
+          last.y = y;
+          last.x = x;
+          return yDelta;
+        }),
+      };
+    },
+  integrate:
+    (unit: keyof typeof timeUnits = "h") =>
+    ({ xs, ys, meta }) => {
+      let yAcc = 0;
+      let last = {
+        x: NaN,
+      };
+      return {
+        meta: {
+          unit_of_measurement: `${meta.unit_of_measurement}*${unit}`,
```

Comment:
I think this should be just the two units concatenated because kW*h is always denoted kWh. Or maybe some thin unbreakable space. Or if you really want something visible, a centered dot like ·

Reply:
Absolutely!